burn-core/src/optim/rmsprop.rs

```diff
@@ -0,0 +1,500 @@
+        self
+    }
+}
+
+/// [RMSPropMomentum](RMSPropMomentum) is to store config status for optimizer.
+/// (, which is stored in [optimizer](RMSProp) itself and not passed in during `step()` calculation)
+pub struct RMSPropMomentum {
+    momentum: f32,
+    epsilon: f32,
+}
+
+impl RMSPropMomentum {
+    /// transform [grad](Tensor) and [RMSPropMomentumState] to the next step
+    pub fn transform<B: Backend, const D: usize>(
+        &self,
+        grad: Tensor<B, D>,
+        centered_state: CenteredState<B, D>,
+        momentum_state: Option<RMSPropMomentumState<B, D>>,
+    ) -> (
+        Tensor<B, D>,
+        CenteredState<B, D>,
+        RMSPropMomentumState<B, D>,
+    ) {
+        let grad = grad
+            .clone()
+            .div(centered_state.avg.clone().sqrt().add_scalar(self.epsilon));
+
+        if self.momentum > 0. {
+            let buf = match momentum_state {
+                Some(state) => state
+                    .buf
+                    .clone()
+                    .mul_scalar(self.momentum)
+                    .add(grad.clone()),
+                _ => grad.clone(),
+            };
+            (buf.clone(), centered_state, RMSPropMomentumState { buf })
+        } else {
+            match momentum_state {
+                Some(state) => (grad.clone(), centered_state, state),
+                _ => (
+                    grad.clone(),
+                    centered_state,
+                    RMSPropMomentumState {
+                        buf: Tensor::zeros(grad.shape()),
+                    },
+                ),
+            }
+        }
+    }
+}
+
+/// [RMSPropMomentumState](RMSPropMomentumState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct RMSPropMomentumState<B: Backend, const D: usize> {
+    buf: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> RMSPropMomentumState<B, D> {
+    fn to_device(mut self, device: &B::Device) -> Self {
+        self.buf = self.buf.to_device(device);
+        self
+    }
+}
+
+impl<B: Backend> SimpleOptimizer<B> for RMSProp<B> {
+    type State<const D: usize> = RMSPropState<B, D>;
+
+    fn step<const D: usize>(
+        &self,
+        lr: LearningRate,
+        mut tensor: Tensor<B, D>,
+        mut grad: Tensor<B, D>,
+        state: Option<Self::State<D>>,
+    ) -> (Tensor<B, D>, Option<Self::State<D>>) {
+        // fetch state for params
+        let mut state_square_avg = None;
+        let mut state_centered = None;
+        let mut state_momentum = None;
+        if let Some(state) = state {
+            state_square_avg = Some(state.square_avg);
+            state_centered = Some(state.centered);
+            state_momentum = Some(state.momentum);
+        }
+
+        // weight_decay transform
+        if let Some(weight_decay) = &self.weight_decay {
+            let (grad_out, tensor_out) = weight_decay.transform_temp_fix(grad, tensor);
+            grad = grad_out;
+            tensor = tensor_out;
```

Comment:
Looking at `transform_temp_fix` I think `tensor_out` is actually just `tensor`, so we don't need to set it here?

Reply:
If it's the case, I would only return the updated gradient from the function `transform_temp_fix`, not a tuple.

Reply:
Emmm, but I need to pass a tensor's reference to this func then. I would thought to keep it looks like the other transform, just move in and move out.

But how about this style?

```rust
if let Some(weight_decay) = &self.weight_decay {
            (grad, tensor) = weight_decay.transform_temp_fix(grad, tensor);
        }
```
still move in, and move out values, but a bit simpler.

Reply:
Done, only grad returned makes it simpler! 😄